File: speech_to_text/nodejs/real_time/stream_radio_translation.js
import fetch from "node-fetch";

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
const apiKey = process.env.SONIOX_API_KEY;
const websocketUrl = "wss://stt-rt.soniox.com/transcribe-websocket";
const audioUrl = "https://npr-ice.streamguys1.com/live.mp3?ck=1742897559135";

// Connect to WebSocket API
const ws = new WebSocket(websocketUrl);

console.log("Opening WebSocket connection...");
ws.addEventListener("open", async () => {
  // Send start request
  ws.send(
    JSON.stringify({
      api_key: apiKey,
      audio_format: "auto", // server detects the format
      model: "stt-rt-preview",
      language_hints: ["en", "es"],
      enable_speaker_diarization: true,
      translation: {
        type: "one_way",
        target_language: "es",
        source_languages: ["en"],
      },
    })
  );

  console.log("Transcription started.");

  try {
    // Connect to the radio stream
    const res = await fetch(audioUrl);
    if (!res.ok) {
      throw new Error(`HTTP error! Status: ${res.status}`);
    }
    res.body.on("data", (chunk) => {
      ws.send(chunk);
    });
    res.body.on("end", () => {
      console.log("Stream ended");
      ws.send(""); // signal end of stream
    });
    res.body.on("error", (error) => {
      console.error("Stream error:", error);
      ws.close();
    });
  } catch (error) {
    console.error("Error sending audio:", error);
    ws.close();
  }
});

function renderTokens(finalTokens, nonFinalTokens) {
  // Render the tokens in the terminal using ANSI escape codes.
  let text = "";
  text += "\x1b[2J\x1b[H"; // Clear the screen and move to the top-left corner
  let isFinal = true;
  let speaker = "";
  let language = "";
  for (const token of finalTokens.concat(nonFinalTokens)) {
    let tokenText = token.text;
    if (!token.is_final && isFinal) {
      text += "\x1b[34m"; // change text color to blue
      isFinal = false;
    }
    if (token.speaker && token.speaker !== speaker) {
      if (speaker) {
        text += "\n\n";
      }
      speaker = token.speaker;
      text += `Speaker ${speaker}: `;
      tokenText = tokenText.trimStart();
      language = "";
    }
    if (token.language && token.language != language) {
      text += "\n";
      language = token.language;
      text += `[${language}] `;
      tokenText = tokenText.trimStart();
    }
    text += tokenText;
  }
  text += "\x1b[39m"; // reset text color
  console.log(text);
}

let finalTokens = [];

ws.addEventListener("message", (event) => {
  // Receive and process messages
  const res = JSON.parse(event.data);

  if (res.error_code) {
    console.log(`\nError: ${res.error_code} ${res.error_message}`);
    process.exit(1);
  }

  let nonFinalTokens = [];

  for (const token of res.tokens || []) {
    if (token.text) {
      if (token.is_final) {
        finalTokens.push(token);
      } else {
        nonFinalTokens.push(token);
      }
    }
  }

  renderTokens(finalTokens, nonFinalTokens);

  if (res.finished) {
    console.log("\nTranscription done.");
  }
});

ws.addEventListener("close", () => {});

ws.addEventListener("error", (error) => {
  console.error("Connection error occurred:", error);
});

File: speech_to_text/python/real_time/stream_radio_translation.py
import json
import os
import threading

import requests
from websockets import ConnectionClosedOK
from websockets.sync.client import connect

# Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
api_key = os.environ.get("SONIOX_API_KEY")
websocket_url = "wss://stt-rt.soniox.com/transcribe-websocket"
audio_url = "https://npr-ice.streamguys1.com/live.mp3?ck=1742897559135"


def stream_audio(ws):
    with requests.get(audio_url, stream=True) as res:
        res.raise_for_status()
        for chunk in res.iter_content(chunk_size=4096):
            if chunk:
                ws.send(chunk)
    ws.send("")  # signal end of stream


def render_tokens(final_tokens, non_final_tokens):
    # Render the tokens in the terminal using ANSI escape codes.
    text = ""
    text += "\033[2J\033[H"  # clear the screen, move to top-left corner
    is_final = True
    speaker = ""
    language = ""
    for token in final_tokens + non_final_tokens:
        token_text = token["text"]
        if not token["is_final"] and is_final:
            text += "\033[34m"  # change text color to blue
            is_final = False
        if token.get("speaker") and token["speaker"] != speaker:
            if speaker:
                text += "\n\n"
            speaker = token["speaker"]
            text += f"Speaker {speaker}: "
            token_text = token_text.lstrip()
            language = ""
        if token.get("language") and token["language"] != language:
            text += "\n"
            language = token["language"]
            text += f"[{language}] "
            token_text = token_text.lstrip()
        text += token_text
    text += "\033[39m"  # reset text color
    print(text)


def main():
    print("Opening WebSocket connection...")

    with connect(websocket_url) as ws:
        # Send start request
        ws.send(
            json.dumps(
                {
                    "api_key": api_key,
                    "audio_format": "auto",  # let the API detect the format
                    "model": "stt-rt-preview-v2",
                    "language_hints": ["en", "es"],
                    "enable_speaker_diarization": True,
                    "translation": {
                        "type": "one_way",
                        "target_language": "es",
                    },
                }
            )
        )

        # Start streaming audio in background
        threading.Thread(target=stream_audio, args=(ws,), daemon=True).start()

        print(f"Transcription started from {audio_url}")

        final_tokens = []

        try:
            while True:
                message = ws.recv()
                res = json.loads(message)

                if res.get("error_code"):
                    print(f"Error: {res['error_code']} - {res['error_message']}")
                    break

                non_final_tokens = []

                for token in res.get("tokens", []):
                    if token.get("text"):
                        if token.get("is_final"):
                            final_tokens.append(token)
                        else:
                            non_final_tokens.append(token)

                render_tokens(final_tokens, non_final_tokens)

                if res.get("finished"):
                    print("\nTranscription complete.")
        except ConnectionClosedOK:
            pass
        except Exception as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    main()
